### PR TITLE
[operator-trivy][admission-policy-engine] Deny the start of the container if High or Critical CVEs are detected in the image

### LINKED ISSUE #3965: Deny the start of the container if High or Critical CVEs are detected in the image
### Preflight Checklist

- [X] I agree to follow the [Code of Conduct](https://github.com/deckhouse/deckhouse/blob/main/CODE_OF_CONDUCT.md) that this project adheres to.
- [X] I have searched the [issue tracker](https://github.com/deckhouse/deckhouse/issues) for an issue that matches the one I want to file, without success.

### Use case. Why is this important?

We've [realized](https://github.com/deckhouse/deckhouse/pull/3858) operator-trivy module that constantly scans container images for CVEs. But it doesn't prevent containers from being launched with CVEs detected though, which is not a security best practice.

### Proposed Solution

Deny the start of the container if High or Critical CVEs are detected in the image + log the event + alert the administrator.

### Additional Information

_No response_

### PR DESCRIPTION
**Pull request by alex123012**
## Description

Deny the start of the container if High or Critical CVEs are detected in the image.

## Why do we need it, and what problem does it solve?

Fix #3965

## What is the expected result?

Create an external data provider for OPA Gatekeeper. Launch trivy-operator in client-server mode. Update trivy and trivy-operator

## Checklist
- [x] The code is covered by unit tests.
- [x] e2e tests passed.
- [x] Documentation updated according to the changes.
- [x] Changes were tested in the Kubernetes cluster manually.

## Changelog entries

```changes
---
section: operator-trivy
type: feature
summary: Run operator-trivy in client-server mode. Update trivy and trivy-operator
impact_level: default
---
section: admission-policy-engine
type: feature
summary: Add an external-data trivy-provider for the gatekeeper to prevent the container from starting if high or critical CVEs are found in the image.
impact_level: default
---
section: tools
type: chore
summary: Add exclude regex for crds/native folder for modules.
```

### CODE REVIEW

**Comment by alex123012** [comment]
### Module configs
```yaml
---
apiVersion: deckhouse.io/v1alpha1
kind: ModuleConfig
metadata:
  name: operator-trivy
spec:
  enabled: true
---
apiVersion: deckhouse.io/v1alpha1
kind: ModuleConfig
metadata:
  name: admission-policy-engine
spec:
  enabled: true
  settings:
    trivyProvider:
      enable: true
  version: 1
```
### Testcases
#### Create
* Not vulnerable `Deployment` and vulnerable `Pod`:
```yaml
---
apiVersion: v1
kind: Namespace
metadata:
  name: test
---
apiVersion: apps/v1
kind: Deployment
metadata:
  name: not-vulnerable-deployment
  namespace: test
spec:
  replicas: 1
  selector:
    matchLabels:
      app: not-vulnerable-deployment
  template:
    metadata:
      labels:
        app: not-vulnerable-deployment
    spec:
      containers:
      - name: not-vulnerable
        image: dev-registry.deckhouse.io/sys/deckhouse-oss/release-channel:alpha
---
apiVersion: v1
kind: Pod
metadata:
  name: vulnerable-pod
  namespace: test
spec:
  containers:
    - name: vulnerable
      image: alpine:3.10
```
image

* Not vulnerable `Pod` and vulnerable `Deployment`:
```yaml
---
apiVersion: v1
kind: Namespace
metadata:
  name: test
---
apiVersion: apps/v1
kind: Deployment
metadata:
  name: vulnerable-deployment
  namespace: test
spec:
  replicas: 1
  selector:
    matchLabels:
      app: vulnerable-deployment
  template:
    metadata:
      labels:
        app: vulnerable-deployment
    spec:
      containers:
      - name: vulnerable
        image: alpine:3.10
---
apiVersion: v1
kind: Pod
metadata:
  name: not-vulnerable-pod
  namespace: test
spec:
  containers:
    - name: not-vulnerable
      image: dev-registry.deckhouse.io/sys/deckhouse-oss/release-channel:alpha
```
image

#### Update
* Not vulnerable `Pod` to vulnerable `Pod` and vulnerable `Deployment` to not vulnerable `Deployment`:
```yaml
---
apiVersion: v1
kind: Namespace
metadata:
  name: test
---
apiVersion: apps/v1
kind: Deployment
metadata:
  name: vulnerable-deployment
  namespace: test
spec:
  replicas: 1
  selector:
    matchLabels:
      app: vulnerable-deployment
  template:
    metadata:
      labels:
        app: vulnerable-deployment
    spec:
      containers:
      - name: vulnerable
        image: dev-registry.deckhouse.io/sys/deckhouse-oss/release-channel:alpha
---
apiVersion: v1
kind: Pod
metadata:
  name: not-vulnerable-pod
  namespace: test
spec:
  containers:
    - name: not-vulnerable
      image: alpine:3.10
```

image

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @alex123012. I've started the workflow for label 'e2e/run/aws'...

File: modules/015-admission-policy-engine/openapi/config-values.yaml [hunk 1]
```diff
@@ -36,3 +36,14 @@ properties:
...
                       type: integer
 
+  trivyProvider:
+    type: object
+    default: {}
+    description: |
+      Trivy provider settings.
+      Trivy provider will deny creation of the `Pod`/`Deployment`/`StatefulSet`/`DaemonSet` with vulnerable images in namespaces with `security.deckhouse.io/trivy-provider: ""` label.
+    properties:
+      enable:
```
Comment (yalosev) [inline modules/015-admission-policy-engine/openapi/config-values.yaml:46]: Also have to be `enabled`

File: modules/015-admission-policy-engine/openapi/config-values.yaml [hunk 2]
```diff
@@ -36,3 +36,14 @@ properties:
                     max:
                       type: integer
 
+  trivyProvider:
```
[the remark this answers is not in the document]
Comment (alex123012) [reply modules/015-admission-policy-engine/openapi/config-values.yaml]: decided to use `denyVulnerableImages`

File: modules/015-admission-policy-engine/charts/constraint-templates/templates/security/vulnerable-images.yaml [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+          response := external_data({"provider": "trivy-provider", "keys": images})
+
+          response_deny_request(response)
+          msg := sprintf("request denied: %v", [response])
+        }
+
+        retrieve_images[imgs] {
+          count(input.review.object.spec.containers[_].image) > 0
+          imgs := input.review.object.spec.containers[_].image
+        }
```
Comment (yalosev) [inline modules/015-admission-policy-engine/charts/constraint-templates/templates/security/vulnerable-images.yaml:31]:
```suggestion
        }
        
        retrieve_images[imgs] {
          count(input.review.object.spec.initContainers[_].image) > 0
          imgs := input.review.object.spec.initContainers[_].image
        }
```

File: modules/015-admission-policy-engine/charts/constraint-templates/templates/security/vulnerable-images.yaml [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+
+        retrieve_images[imgs] {
+          count(input.review.object.spec.template.spec.containers) > 0
+          imgs := input.review.object.spec.template.spec.containers[_].image
+        }
```
Comment (yalosev) [inline modules/015-admission-policy-engine/charts/constraint-templates/templates/security/vulnerable-images.yaml:36]:
```suggestion
        }
        
        retrieve_images[imgs] {
          count(input.review.object.spec.template.spec. initContainers) > 0
          imgs := input.review.object.spec.template.spec. initContainers[_].image
        }
```

File: modules/015-admission-policy-engine/openapi/config-values.yaml [hunk 5]
```diff
@@ -36,3 +36,28 @@ properties:
...
+    properties:
+      enabled:
+        type: boolean
+        default: false
+        description: "enable trivy provider (EE only)"
+      registrySecrets:
+        type: array
+        default: []
+        description: |
+          Registry secrets to use for downloading images from private registries.
```
Comment (yalosev) [inline modules/015-admission-policy-engine/openapi/config-values.yaml:53]:
```suggestion
          Additional registry secrets to use for downloading images from private registries. By default, only `deckhouse-registry` secret is used. If you store all images in the same registry, you don't have to pass more secrets here.
```

**Review by yalosev — APPROVED** [review]

**Review by nabokihms — APPROVED** [review]

**Pull request merged**